apps/state/lib/state/vehicle.ex

```diff
@@ -101,7 +99,32 @@ defmodule State.Vehicle do
     end
   end
 
-  defp build_filters(matchers, key, values, _filters) do
-    for matcher <- matchers, value <- values, do: Map.put(matcher, key, value)
+  @spec do_post_search_filter([Vehicle.t()], filter_opts) :: [Vehicle.t()]
+  defp do_post_search_filter(vehicles, %{labels: labels}) when is_list(labels) do
+    labels = MapSet.new(labels)
+
+    consist_matches? = fn %Model.Vehicle{consist: consist}, labels ->
+      case consist do
+        nil ->
+          false
+
+        _ ->
+          consist
+          |> MapSet.new()
+          |> MapSet.intersection(labels)
+          |> MapSet.size()
+          |> Kernel.!==(0)
+      end
+    end
+
+    label_matches? = fn %Model.Vehicle{label: label}, labels ->
+      label in labels
+    end
+
+    Enum.filter(vehicles, fn vehicle ->
+      label_matches?.(vehicle, labels) or consist_matches?.(vehicle, labels)
+    end)
```

Comment:
Since you're building the match functions here anyways, you don't need to pass in `labels`. It's available from the higher scope.